### ISSUE TITLE
segmentation fault when not started with -S

### ISSUE THREAD

**Issue by muellerto** [leo-arch/CliFM#300]
**Describe the bug**
clifm produces a segmentation fault when I don't use `-S` at startup. But my configuration files are completely default after first time installation.

**To Reproduce**
I did a complete reinstall of my msys2 environment including MinGW64. Then I got actual sources from the Git repository and did a fresh rebuild of clifm. I used `make -f misc/cygwin/Makefile install` to copy all the files to my `/usr/local` directories. But then, the first start as just `clifm` produces a segmentation fault. You see nothing, not a flicker. The shell command prompt returns very fast, in under a second.

Before the start I didn't have a local `.config/clifm` directory. After the crash I have one but there are only two files in it:
```
.config/clifm/profiles/default/clifmrc           <- looks good
.config/clifm/profiles/default/msglogs.clifm     <- 0 bytes!
```
The segmentation fault doesn't come when I give him a `-S` at startup.

`clifm --help` does not crash and shows the help.

**Expected behavior**
It should at least say what's not good :(

**Desktop (please complete the following information):**
- OS:Windows: 11 Pro
- Terminal: Hyper (doesn't matter)
- CliFM version: 1.20-rc2 (Aug 13, 2024)
- Installation source: Git from today

**Additional context**
I tried to switch on logging in `clifmrc`. But I guess it crashes before it reads this setting.

**Comment by leo-arch** [comment]
And I was planning to release 1.20 today. I guess it won't happen.

Ok, just a quick test: running the latest clifm with `-D ~/.config/clifm.test` to start with a new configuration directory. I don't get a crash, but a bunch of permission denied messages about each new configuration file. However, files are actually created and they all have the appropriate permissions (644)! Indeed, everything is fine, no warnings, in the next run.

What we know up to now is this:

1. We have some (permissions?) issue when creating config files under Windows (actually, copying them from the data directory).
2. The crash, though not reproduced, might be triggered when trying to log one of these warnings (too bad I haven't implemented a `--no-log` option yet!).

For the time being, I guess you can work around this issue by manually copying the config files from the data directory (/usr/local/share/clifm).

I'll investigate this issue further.

**Comment by muellerto** [comment]
Make a release :) As long as only I have this on Windows the whole problem is rather exotic. I also don't have this on Linux.

The permissions should be OK. Was also my first idea. But I can write a file there and we saw clifm can also do it.

What I meanwhile saw is:
- it crashes also with a complete clean and sober `.config/clifm` directory from another machine which then shouldn't been written, only read.
- when I start with `-S -s` it shows the splash screen and then terminates
 
Tomorrow I check this `--no-log` switch. Good idea.

I could also debug this. GDB runs fine in Visual Studio Code ...

**Comment by leo-arch** [comment]
> Tomorrow I check this --no-log switch. Good idea.

Don't, it doesn't exist (but would be useful now).

> I could also debug this. GDB runs fine in Visual Studio Code ...

This would be ideal.

**Comment by leo-arch** [comment]
Found [this](https://github.com/msys2/MSYS2-packages/issues/2612): chmod does not work on MSYS2. Tested it on Cygwin: works as expected. For the time being I'll add a conditional macro to not use chmod on MSYS2.

EDIT: Done (https://github.com/leo-arch/clifm/commit/d03e3adba5f892a7a3e2d1371b67acb4d59206da). Btw, can you try compiling on Cygwin; it might be an MSYS2-only issue.

**Comment by muellerto** [comment]
My gdb session showed the following: it crashes in `truncate_file()` in line 935:
```
	while (getline(&line, &line_size, orig_fp) > 0) {
		/* Skip consecutive equal entries. */
		if (check_dups == 1 && prev_line && line_size == prev_line_size
		&& strcmp(line, prev_line) == 0)
			continue;

		/* Delete old entries, i.e., copy only new ones. */
		if (i++ >= n - (max - 1))
			fprintf(tmp_fp, "%s", line);                 <------- line 935

		if (check_dups == 1) {
			free(prev_line);
			prev_line = savestring(line, line_size);
			prev_line_size = line_size;
		}
	}
```
`line` has a proper value and a 0 byte at the end. The file is a temporary one, it exists, it's not read only and I have (as owner) full access, so it should be writable without any problem.

It is the first time that this line 935 is executed. I'm not sure if this is the first `fprintf()` call at all (I guess it is).

I changed the line 935 into a simple `fputs()` call to make a test. This solves this problem immediately. But it crashes later in `append_to_dirhist_file()`, there's the next `fprintf()` call. Seems that the MinGW `fprintf()` implementation is not the best. Could be a workaround to implement an own function for MinGW using `printf()` and then `fputs()`. Yeah, not so easy with a variable list of arguments ....

A question is: why does it not crash when `-S` is specified? Is there really not a single `fprintf()` call? I saw you use it a lot, also for `stderr`. I think it will crash also, just somewhen later.

**Comment by leo-arch** [comment]
Great info! Does GDB say something about the reason why `fprintf` fails?

> Is there really not a single fprintf() call?

I can assure you there are, but maybe not to write stuff into a file, but in memory (a variable). Maybe MinGW implementation of `fprintf` just fails to write data _into a file_.

EDIT: The `fprintf` call in `append_to_dirhist_file()` can be replaced by this:

```c
fputs(dir_path, fp);
fputc('\n', fp);
```
Can you give this a try?

**Comment by muellerto** [comment]
Sure, works fine (and should also be faster than `fprintf()`) but then it crashes later in `create_config_files()`.

No, gdb tells no reason. `fprintf()` is just the last function called directly from clifm on the backtrace (call stack).

You could make some experiments with a mini program just writing a file with `fprintf()` on MinGW. I can't imagine what's the problem with `fprintf()`, this function exists for decades and has been tested in thousands of use cases.

BTW: You must solve this alone now because I'm in holidays for the next two weeks :)

**Comment by muellerto** [comment]
Two more thoughts:
- there was a guy here (from France?) some weeks ago who also did a MinGW compile, ask him if he also has these problems
- when this problem is really related to the `fprintf()` implementation it's a problem of the glibc I use, this is the one from MSYS2, in this case we should open a case there, the issue tracking is [here](https://github.com/msys2/msys2-packages/issues)

**Comment by leo-arch** [comment]
As far as I can see, the problem here is `fprintf`, and thereby, as you said, the C library, in which case there's nothing we can do. What version are you using? My `ldd --version` output is:

```
ldd (cygwin) 3.3.5 # This is Cygwin
ldd (cygwin) 3.4.6 # This is MSYS
```
`fprintf` works fine in both cases.

**Comment by muellerto** [comment]
The longer I think about this the less I'm sure.

The integration of MinGW into MSYS2 is a mysterious thing. I installed MSYS2, not MinGW. The gcc I use is the one which is part of MSYS2, 13.2 when I remember well. You can install a specific MinGW gcc as a package which is then 14.1 or so. I didn't do that (because of other following problems and ambiguities).

But even when compiled with the original gcc from MSYS2 clifm needs some headers which are exclusively located in the `mingw64/include` sub directory, for instance `libintl.h`, libraries from `mingw64/lib` and (may be) DLLs from `minw64/bin`. Because of these needs I had to extend the gcc default search paths for includes and for libraries, otherwise you get both compile and link time errors. I did this in the [specs file of gcc](https://gcc.gnu.org/onlinedocs/gcc/Spec-Files.html) (ignoring some often read warnings not to do that ...), so it compiles and links.

But I'm not sure about runtime. I didn't change `LD_LIBRARY_PATH` until now. A probable case could be that I link with a concrete library but clifm loads then another library for runtime. I must analyze this. The problem is I can't do this in the next two weeks because I'm not on Windows in this time and rather seldom on a machine at all. I close this case until I know more.

BTW: Cygwin, yes, I know. But I avoided Cygwin until today for two reasons: it's rather old (and in some cases even deprecated) and it's also limited. They are proud to have no package manager for installing more (because "this is not what you want"), so it's just as is.

**Comment by leo-arch** [comment]
Hey @muellerto, no problem. Make all the tests you need whenever you can. I'll be happy to help.

Btw, the `libintl.h` header is only for `gettext`, which is for translations. If you don't need this, as is probably the case, you can compile with `-D_NO_GETTEXT`, and remove `-lintl` from the `gcc` command.

**Comment by muellerto** [comment]
Now after my holidays I removed my selfmade specs file again to restart under clear conditions. The entire build environment should exactly be as it comes from MSYS2. But I also got some MSYS2 updates using pacman, so it isn't exactly the same anymore as it has been two weeks ago.

What should I say ... now clifm compiles without a problem and it also runs!

Damned. I can't explain it. I can't say anymore if I really had only a bad moment in the MSYS2 distribution or if it was another reason. It seems to be solved by itself.
